Revise migration guide from Statsig to Datadog (#34531)
DataDog/documentation: hugo/content/en/feature_flags/guide/migrate_from_statsig.md

---
title: Migrate Your Feature Flags from Statsig
description: Learn how to migrate feature flags from Statsig to Eppo by Datadog.
description: Learn how to migrate feature flags from Statsig to Datadog.
---

## Overview
This guide outlines the process for migrating your feature flagging logic from Statsig to [Datadog Feature Flags][1]. It covers conceptual mappings, SDK installation, initialization, and flag evaluation.

This guide walks you through the process of migrating feature flags from Statsig to [Eppo by Datadog][1], as an intermediate step before fully migrating to Datadog's dedicated [Feature Flags][2] product. Follow these general steps:
## Summary checklist

1. [Install the Eppo SDK.](#install-sdk)
2. [Create a feature flag in Eppo and verify its functionality.](#set-up-flag)
3. [Identify critical feature flags in Statsig.](#identify-critical-flags)
4. [For all non-critical flags, remove existing code.](#remove-non-critical-flags)
5. [For critical flags, create a fallback value in a wrapper.](#create-fallback-values)

> 6. [Recreate critical feature flags in Eppo.](#recreate-critical-flags)

Reviewer: We'll still need a step like this--flags will need to be recreated in Datadog.

const config = getInstance().getJSONAssignment(user.userID, 'product_config', userAttributes, {});
const buttonColor = config?.button_color || 'blue';
const maxItems = config?.max_items || 10;

// For individual string parameters:
const buttonColor = getInstance().getStringAssignment(
  user.userID,
  'button_color_flag',
  userAttributes,
  'blue'
);
### Datadog (new)

// For individual numeric parameters:
const maxItems = getInstance().getNumericAssignment(
  user.userID,
  'max_items_flag',
  userAttributes,
  10
);
{{< code-block lang="javascript" >}}
// Update the context for all future flag evaluations
await OpenFeature.setContext({
  targetingKey: 'user-456', // Maps to the user ID in Datadog
  email: 'employee@company.com',
  plan: 'premium'
});
{{< /code-block >}}

[1]: https://docs.geteppo.com/sdks/server-sdks/node/assignments/
{{% /tab %}}
{{< /tabs >}}

### Get experiment values
## 7. Tracking and exposure

For example, get experiment parameter values.
In Statsig, checking a gate automatically logs an exposure.

{{< tabs >}}
{{% tab "Statsig" %}}
Statsig docs: [Getting a Layer/Experiment][1]
{{< code-block lang="typescript" >}}

> // Values using getLayer

Reviewer: Should we explicitly talk about using JSON-valued flags to replace StatSig layers? You imply it in a later example.

| **Layer** | **Feature Flag** (JSON variant) | Use a JSON-valued flag and read fields from the returned object, similar to dereferencing values from a Statsig layer. |
| **Experiment** | **Feature Flag** (with targeting) | A Datadog flag can be configured with percentage-based rollouts and specific targeting rules to run experiments. Connect flags to [Datadog Experiments][5] to measure impact on user outcomes. |
| **User/StatsigUser** | **Evaluation Context** | The context (attributes) passed to the SDK to evaluate flags. |

## Installation

Datadog designs its feature flagging SDKs for use with [OpenFeature][6]. This provides a vendor-neutral API while using Datadog as the underlying provider.

Remove Statsig:

{{< code-block lang="bash" >}}
npm uninstall @statsig/js-client
# or
yarn remove @statsig/js-client
{{< /code-block >}}

Install Datadog and OpenFeature:

{{< code-block lang="bash" >}}
npm install @datadog/openfeature-browser @openfeature/web-sdk @openfeature/core
# or
yarn add @datadog/openfeature-browser @openfeature/web-sdk @openfeature/core
{{< /code-block >}}

> **Note**: For React applications, also install `@openfeature/react-sdk`. See [React Feature Flags][7]. For server-side implementations, see [Server-side and dynamic context](#server-side-and-dynamic-context) below, or the [Datadog Server SDK documentation][2] for other languages.

Reviewer 2: ```suggestion
**Note**: For React applications, also install `@openfeature/react-sdk`. See [React Feature Flags][7]. For server-side implementations, see the [Server-side and dynamic context](#server-side-and-dynamic-context) section, or [Server-Side Feature Flags][2] for other languages.
```

const promoTitle = layer.get('title', 'Welcome to Statsig!');
const discount = layer.get('discount', 0.1);
{{< /code-block >}}

### Datadog (new)

{{< code-block lang="javascript" >}}
const client = OpenFeature.getClient();

const promoConfig = client.getObjectValue('user_promo_experiments', {
  title: 'Welcome!',
  discount: 0.1,
});
const promoTitle = promoConfig.title;
const discount = promoConfig.discount;
{{< /code-block >}}

## Update user context after login

Statsig updates user context using `updateUser`. In OpenFeature and Datadog, update the context after initialization with `OpenFeature.setContext()`, for example after a user logs in.

### Statsig (old)

{{< code-block lang="javascript" >}}
await client.updateUserAsync({
  userID: 'user-456',
  email: 'employee@company.com',
  custom: { plan: 'premium' },
});
{{< /code-block >}}

### Datadog (new)

{{< code-block lang="javascript" >}}
// Update the context for all future flag evaluations
await OpenFeature.setContext({
  targetingKey: 'user-456', // Identifies the user and drives percentage-based randomization
  email: 'employee@company.com',
  plan: 'premium',
});
{{< /code-block >}}

## Tracking and exposure

> In Statsig, checking a gate automatically logs an exposure.

Reviewer 2: could we include somewhere here a definition of exposure logging and evaluation logging, with an explanation of how they differ?

Author: done!

In Datadog:

1. **Client SDKs**: Exposure logging is enabled by default. Exposure events are sent to Datadog RUM. You can view them in the **Feature Flags** list or the **RUM** explorer. Set `enableExposureLogging: false` in the `DatadogProvider` config if you do not need exposure tracking.

<div class="alert alert-warning">Setting <code>enableExposureLogging</code> to <code>true</code> can impact <a href="/real_user_monitoring/">RUM</a> costs, as it sends exposure events to Datadog through RUM. Exposure logging is on by default for client SDKs.</div>

> 2. **Server SDKs**: Exposure logging is on by default. Evaluation logging is off by default. To send evaluation metrics from server SDKs, enable OpenTelemetry metrics (for example, `DD_METRICS_OTEL_ENABLED=true`) and follow the language-specific guidance in the [Datadog Server SDK documentation][2].

Reviewer 2: ```suggestion
2. **Server SDKs**: Exposure logging is on by default. Evaluation logging is off by default. To send evaluation metrics from server SDKs, enable OpenTelemetry metrics (for example, `DD_METRICS_OTEL_ENABLED=true`) and follow the language-specific guidance in [Server-Side Feature Flags][2].
```

{{< /code-block >}}

Register the provider through the Datadog tracer:

{{< code-block lang="javascript" >}}
const tracer = require('dd-trace').init();
const { OpenFeature } = require('@openfeature/server-sdk');

await OpenFeature.setProviderAndWait(tracer.openfeature);
{{< /code-block >}}

### Statsig (old)

{{< code-block lang="javascript" >}}
const isEnabled = client.checkGate('new_homepage_design');
{{< /code-block >}}

### Datadog (new)

{{< code-block lang="javascript" >}}
const client = OpenFeature.getClient();

app.get('/my-endpoint', async (req, res) => {
  const evaluationContext = {
    targetingKey: req.session?.userID ?? 'unknown',
  };

  const isEnabled = await client.getBooleanValue('new_homepage_design', false, evaluationContext);
  res.send(isEnabled ? 'New design' : 'Old design');
});
{{< /code-block >}}

The browser SDK uses whatever evaluation context is set for every flag evaluation. You can update that context with `OpenFeature.setContext()` when the user logs in or their attributes change. The server SDK instead passes a new evaluation context into each flag evaluation call, since one process handles many different users.

> For other server languages, see the [Datadog Server SDK documentation][2].

Reviewer 2: ```suggestion
For other server languages, see [Server-Side Feature Flags][2].
```

---
title: Migrate Your Feature Flags from Statsig

> description: Learn how to migrate feature flags from Statsig to Datadog.

Reviewer: 🤖 **From Claude working with Aaron:** Add "Feature Flags" to the description. This is the product name and helps search.

The LaunchDarkly guide uses this form.

```yaml
description: Learn how to migrate feature flags from Statsig to Datadog Feature Flags.
```

---

> This guide outlines the process for migrating your feature flagging logic from Statsig to [Datadog Feature Flags][1]. It covers conceptual mappings, SDK installation, initialization, and flag evaluation.

Reviewer: 🤖 **From Claude working with Aaron:** Add `further_reading` frontmatter and an `## Overview` heading. All five other guides in this directory have both.

Add to the frontmatter:

```yaml
further_reading:
- link: "/feature_flags/"
  tag: "Documentation"
  text: "Feature Flags Overview"
- link: "/feature_flags/client/"
  tag: "Documentation"
  text: "Client-Side Feature Flags"
- link: "/feature_flags/server/"
  tag: "Documentation"
  text: "Server-Side Feature Flags"
```

Add above the link definitions at the end of the file:

```
{{< partial name="whats-next/whats-next.html" >}}
```

Replace `checkGate` calls with OpenFeature's `getBooleanValue`.

### Statsig (old)

{{< code-block lang="javascript" >}}
const isEnabled = client.checkGate('new_homepage_design');

if (isEnabled) {
  // Show new design
} else {
  // Show old design
}
{{< /code-block >}}

### Datadog (new)

{{< code-block lang="javascript" >}}
const client = OpenFeature.getClient();

// The second argument is the fallback value (default) if the flag fails to fetch
const isEnabled = client.getBooleanValue('new_homepage_design', false);

if (isEnabled) {
  // Show new design
} else {
  // Show old design
}
{{< /code-block >}}

## Get configuration (dynamic configs)

If you were using `getDynamicConfig` or `getExperiment` to retrieve non-Boolean values (strings, JSON, numbers), use the appropriate typed method in OpenFeature.

### Statsig (old)

{{< code-block lang="javascript" >}}
const config = client.getDynamicConfig('banner_config');
const title = config.get('title', 'Welcome');
{{< /code-block >}}

### Datadog (new)

{{< code-block lang="javascript" >}}
const client = OpenFeature.getClient();

// Assuming your Datadog flag 'banner_config' returns a JSON object variant

> const bannerConfig = client.getObjectValue('banner_config', { title: 'Welcome' });

Reviewer: 🤖 **From Claude working with Aaron:** Add a type parameter to `getObjectValue` and set the block language to `typescript`. TypeScript users cannot read fields from the result as written.

`getObjectValue` returns `JsonValue`. Reading `.title` fails to compile:

```
error TS2339: Property 'title' does not exist on type
  'string | number | boolean | JsonArray | JsonObject'.
```

```typescript
const bannerConfig = client.getObjectValue<{ title: string }>('banner_config', {
  title: 'Welcome',
});
const title = bannerConfig.title;
```

The layer example at line 169 needs the same change.

{{< /code-block >}}

### Datadog (new)

{{< code-block lang="javascript" >}}
const client = OpenFeature.getClient();

const promoConfig = client.getObjectValue('user_promo_experiments', {
  title: 'Welcome!',
  discount: 0.1,
});
const promoTitle = promoConfig.title;
const discount = promoConfig.discount;
{{< /code-block >}}

## Update user context after login

Statsig updates user context using `updateUser`. In OpenFeature and Datadog, update the context after initialization with `OpenFeature.setContext()`, for example after a user logs in.

### Statsig (old)

{{< code-block lang="javascript" >}}
await client.updateUserAsync({
  userID: 'user-456',
  email: 'employee@company.com',
  custom: { plan: 'premium' },
});
{{< /code-block >}}

### Datadog (new)

{{< code-block lang="javascript" >}}
// Update the context for all future flag evaluations
await OpenFeature.setContext({
  targetingKey: 'user-456', // Identifies the user and drives percentage-based randomization
  email: 'employee@company.com',
  plan: 'premium',
});
{{< /code-block >}}

## Tracking and exposure

In Statsig, checking a gate automatically logs an exposure.

In Datadog, flag telemetry falls into two categories:

> **Exposure logging** records that a subject received a specific flag variant, variant served, and evaluation served. Use exposure data to analyze experiment results and feature adoption.

Reviewer: 🤖 **From Claude working with Aaron:** Correct this sentence. It repeats "variant served" and states "evaluation served".

An exposure event contains the flag key, the variant served, and the evaluation context.

```markdown
**Exposure logging** records that a subject received a specific flag variant. Each exposure event includes the flag key, variant served, and evaluation context. Use exposure data to analyze experiment results and feature adoption.
```

> **Exposure logging** records that a subject received a specific flag variant. Each exposure event includes the flag key, variant served, and evaluation context. Use exposure data to analyze experiment results and feature adoption.

Reviewer 2: ```suggestion
**Exposure logging** records that a subject received a specific flag variant, variant served, and evaluation served. Use exposure data to analyze experiment results and feature adoption.
```

Reviewer: 🤖 **From Claude working with Aaron:** This suggestion was applied, but it introduced an error. The text now repeats "variant served" and states "evaluation served". I put the correction on line 208.

> **Evaluation logging** records how often each variant is returned. Client SDKs send aggregated evaluation telemetry by default. Server SDKs emit the `feature_flag.evaluations` metric when evaluation logging is enabled.

Reviewer 2: ```suggestion
**Evaluation logging** records how often each variant is returned. Client SDKs send aggregated evaluation counts by default. Server SDKs emit the `feature_flag.evaluations` metric only after you enable evaluation logging.
```

> Exposure logging tracks which variant a subject saw. Evaluation logging measures how frequently each variant is returned across your application.

Reviewer: 🤖 **From Claude working with Aaron:** Delete this line. Lines 208 and 210 already define both terms.

> 1. **Client SDKs**: Exposure logging is enabled by default. Exposure events are sent to Datadog RUM. You can view them in the **Feature Flags** list or the **RUM** explorer. Set `enableExposureLogging: false` in the `DatadogProvider` config if you do not need exposure tracking.

Reviewer: 🤖 **From Claude working with Aaron:** Change "sent to Datadog RUM" to the exposures intake. Client SDKs do not send exposures through RUM.

```markdown
1. **Client SDKs**: Exposure logging is enabled by default. The SDK sends exposure events to the exposures intake. You can view them in the **Feature Flags** list. Set `enableExposureLogging: false` in the `DatadogProvider` config if you do not need exposure tracking.
```

> <div class="alert alert-warning">Setting <code>enableExposureLogging</code> to <code>true</code> can impact <a href="/real_user_monitoring/">RUM</a> costs, as it sends exposure events to Datadog through RUM. Exposure logging is on by default for client SDKs.</div>

Reviewer: 🤖 **From Claude working with Aaron:** Name `enableRumFeatureFlagTracking` in this alert. `enableExposureLogging` does not affect RUM costs.

`enableExposureLogging` sends events to the exposures intake. `enableRumFeatureFlagTracking` adds flags to RUM events. Only that option increases RUM billing. See [Configure browser provider options][a].

The alert also says "setting to `true`". Both options default to `true`.

```html
<div class="alert alert-warning">The <code>enableRumFeatureFlagTracking</code> option (default <code>true</code>) adds flag evaluations to RUM events. This can increase <a href="/real_user_monitoring/">RUM</a>-billed event counts. Set it to <code>false</code> to stop this.</div>
```

[a]: /feature_flags/client/javascript/#configure-browser-provider-options

2. **Server SDKs**: Exposure logging is on by default. Evaluation logging is off by default. To send evaluation metrics from server SDKs, enable OpenTelemetry metrics (for example, `DD_METRICS_OTEL_ENABLED=true`) and follow the language-specific guidance in [Server-Side Feature Flags][2].

## Server-side and dynamic context {#server-side-and-dynamic-context}

The previous sections cover browser and client-side migration, where the evaluation context is typically static for the length of a user's session. Server-side applications use a different SDK and authenticate with a Datadog API key instead of a client token. They also typically build a new evaluation context for each incoming request.

Configure the required environment variables before initializing the server SDK:

{{< code-block lang="bash" >}}
DD_API_KEY=<DATADOG_API_KEY>
DD_SITE=<DATADOG_SITE>
DD_ENV=<ENVIRONMENT_NAME>
DD_SERVICE=<YOUR_SERVICE_NAME>
DD_REMOTE_CONFIG_ENABLED=true

> DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true

Reviewer: 🤖 **From Claude working with Aaron:** Remove `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED` and `DD_REMOTE_CONFIG_ENABLED`. Agentless delivery is the default and needs three variables.

[Set Up Server-Side Flag Evaluation Metrics][a] states: "Do not set `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true` ... The setting is deprecated."

`DD_REMOTE_CONFIG_ENABLED` controls Agent Remote Configuration. Agentless delivery does not use it.

```bash
DD_API_KEY=<DATADOG_API_KEY>
DD_SITE={{< region-param key="dd_site" code="true" >}}
DD_ENV=<ENVIRONMENT_NAME>
DD_SERVICE=<YOUR_SERVICE_NAME>
```

The `region-param` shortcode also renders the correct site per region.

[a]: /feature_flags/guide/server_flag_evaluation_metrics/

{{< /code-block >}}

See [Server-Side Feature Flags][2] for the full list of Agent and application configuration options.

Install the server-side SDK. This example uses the [Node.js Feature Flags SDK][3]:

{{< code-block lang="bash" >}}
npm install dd-trace @openfeature/server-sdk
{{< /code-block >}}

Register the provider through the Datadog tracer:

{{< code-block lang="javascript" >}}
const tracer = require('dd-trace').init();
const { OpenFeature } = require('@openfeature/server-sdk');

> await OpenFeature.setProviderAndWait(tracer.openfeature);

Reviewer: 🤖 **From Claude working with Aaron:** Change this block to ESM. The snippet does not run as written.

`require()` makes the file CommonJS. Top-level `await` is illegal in CommonJS. `node --check` reports:

```
SyntaxError: await is only valid in async functions and the top level bodies of modules
```

[Node.js Feature Flags][a] uses ESM for this reason.

```javascript
import { OpenFeature } from '@openfeature/server-sdk';
import tracer from 'dd-trace';

tracer.init();

await OpenFeature.setProviderAndWait(tracer.openfeature);
```

To keep `require()`, wrap the `await` in an `async` function.

[a]: /feature_flags/server/nodejs/

{{< /code-block >}}

### Statsig (old)

{{< code-block lang="javascript" >}}

> const isEnabled = client.checkGate('new_homepage_design');

Reviewer: 🤖 **From Claude working with Aaron:** Use the Statsig server signature here. This line shows the browser API.

Statsig server SDKs pass the user in each call. This matches the Datadog server example below. It also shows the reader what actually changes.

```javascript
// The Statsig server SDK takes the user in each call
const isEnabled = statsig.checkGate(user, 'new_homepage_design');
```

{{< /code-block >}}

### Datadog (new)

{{< code-block lang="javascript" >}}
const client = OpenFeature.getClient();

app.get('/my-endpoint', async (req, res) => {
  const evaluationContext = {
    targetingKey: req.session?.userID ?? 'unknown',
  };

  const isEnabled = await client.getBooleanValue('new_homepage_design', false, evaluationContext);
  res.send(isEnabled ? 'New design' : 'Old design');
});
{{< /code-block >}}

The browser SDK uses whatever evaluation context is set for every flag evaluation. You can update that context with `OpenFeature.setContext()` when the user logs in or their attributes change. The server SDK instead passes a new evaluation context into each flag evaluation call, since one process handles many different users.

For other server languages, see [Server-Side Feature Flags][2].

[1]: /feature_flags/
[2]: /feature_flags/server/
[3]: /feature_flags/server/nodejs/
[4]: /account_management/api-app-keys/

> [5]: /product_analytics/experimentation/

Reviewer: 🤖 **From Claude working with Aaron:** Change link `[5]` to `/experiments/`. The current target does not exist.

`/product_analytics/experimentation/` returns a 404. The other Feature Flags guides use `/experiments/`.

```markdown
[5]: /experiments/
```
